### PR TITLE
1.11: more methodinstance stuff

### PR DESCRIPTION
**Pull request by wsmoses**

### CODE REVIEW

File: src/utils.jl [hunk 1]
```diff
@@ -253,24 +253,73 @@ export codegen_world_age
...
-    mi === nothing && throw(MethodError(ft, tt, world))
-    mi = mi::MethodInstance
-
-    # `jl_method_lookup_by_tt` and `jl_method_lookup` can return a unspecialized mi
-    if !Base.isdispatchtuple(mi.specTypes)
-        mi = Core.Compiler.specialize_method(mi.def, sig, mi.sparam_vals)::MethodInstance
+    if Base.isdispatchtuple(sig)   # JuliaLang/julia#52233
+        return GPUCompiler.methodinstance(ft, tt, world)
+    else
+        return prevmethodinstance(ft, tt, world)
```
Comment (vchuravy) [inline src/utils.jl:321]: You just want to call `mi = Core.Compiler.specialize_method(mi.def, sig, mi.sparam_vals)::MethodInstance` in this case, but you need a call to which to obtain `def`

Comment (vchuravy) [reply src/utils.jl:321]: Edit: Maybe? This is all pretty much unsound.

Comment (wsmoses) [reply src/utils.jl:321]: I mean happy to defer that to the experts, it's now just falling back to the old implementation

Comment (wsmoses) [reply src/utils.jl:321]: Old from 1.10 in gpucompiler*** while concurrently I opened a PR to gpucompiler jl to actually fix and not have us have to vendor it

Comment (wsmoses) [reply src/utils.jl:321]: https://github.com/JuliaGPU/GPUCompiler.jl/pull/644

**Pull request merged**